### PR TITLE
Calibration curve

### LINKED ISSUE #118: Better tracker integration Calibration curve
See the issue below, apply for the titled plot
[Reference](https://github.com/ploomber/sklearn-evaluation/issues/84)

### PR DESCRIPTION
**Pull request by neelasha23**
## Describe your changes
Calibration curve objected-oriented API

## Issue ticket number and link
Closes #118 

## Checklist before requesting a review

- [x] I have performed a self-review of my code
- [x] I have added thorough [tests](https://github.com/ploomber/contributing/blob/main/CONTRIBUTING.md#testing) (when necessary).
- [x] I have added the right documentation in the [docstring](https://github.com/ploomber/contributing/blob/main/CONTRIBUTING.md#documenting-changes-and-new-features) and [changelog](https://github.com/ploomber/contributing/blob/main/CONTRIBUTING.md#changelog) (when needed)

----
:books: Documentation preview :books:: https://sklearn-evaluation--265.org.readthedocs.build/en/265/

### CODE REVIEW

**Comment by neelasha23** [comment]
Sure. The CI is also failing, so will tag once it's sorted out.

**Comment by edublancas** [comment]
ci has been fixed in master, please rebase

File: CHANGELOG.md [hunk 1]
```diff
@@ -4,6 +4,9 @@
 
 ## 0.11.0 (2023-01-31)
 
+- [API Change] Adds `plot.CalibrationCurve` object-oriented API
+- [Doc] User Guide on Probability Calibration
```
Comment (edublancas) [inline CHANGELOG.md:8]: changelog notes should always go in the top section (development version), please update

File: docs/api/plot.rst [hunk 2]
```diff
@@ -119,10 +124,6 @@ roc
 ---
 .. autofunction:: sklearn_evaluation.plot.roc
 
-scores_distribution
--------------------
-.. autofunction:: sklearn_evaluation.plot.scores_distribution
-
```
Comment (edublancas) [inline docs/api/plot.rst:125]: looks like this was removed by accident?

File: docs/classification/calibration.md [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+
+The better calibrated or more reliable a forecast, the closer the points will appear along the main diagonal from the bottom left to the top right of the plot.
+
+Calibration of prediction probabilities is a rescaling operation that is applied after the predictions have been made by a predictive model.
+
+Calibration of a classifier can be checked by creating a reliability diagram as demonstrated below.
+
++++
+
+## Reliability Diagrams
```
Comment (edublancas) [inline docs/classification/calibration.md:28]: let's call this calibration to be consistent with our plot's name

File: docs/classification/calibration.md [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+```
+
+```{code-cell} ipython3
+from sklearn_evaluation import plot
+
+plot.CalibrationCurve.from_raw_data(y_test, probabilities, clf_names=clf_names)
+```
+
+## Scores distribution
+
```
Comment (edublancas) [inline docs/classification/calibration.md:91]: let's add some context on why the scores distribution can complement the calibration curve, our blog post contains some explanation: https://ploomber.io/blog/calibration-curve/

File: docs/classification/calibration.md [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+LogisticRegression returns well calibrated predictions by default. In contrast, the other methods return biased probabilities.
+
+GaussianNB tends to push probabilities to 0 or 1 (note the counts in the histograms). This is mainly because it makes the assumption that features are conditionally independent given the class, which is not the case in this dataset which contains 2 redundant features.
+
++++
+
+References:
+
+https://scikit-learn.org/stable/modules/calibration.html
```
Comment (edublancas) [inline docs/classification/calibration.md:115]: add the links in .md format

File: docs/classification/calibration.md [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+
+https://machinelearningmastery.com/calibrated-classification-model-in-scikit-learn/
```
Comment (edublancas) [inline docs/classification/calibration.md:117]: replace this with our blog entry: https://ploomber.io/blog/calibration-curve/

File: docs/api/plot.rst (hunk 2 above)
Comment (edublancas) [reply docs/api/plot.rst:125]: I see what happened. you put this once inside CalibrationCurve. let's revert it to how it was before, having two plots in the same class will be confusing

File: docs/classification/calibration.md [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+```{code-cell} ipython3
+from sklearn.pipeline import make_pipeline
+from sklearn.preprocessing import StandardScaler
+from sklearn.svm import SVC
+
+svc = make_pipeline(StandardScaler(), SVC(gamma='auto', probability=True))
+svc_probas = svc.fit(X_train, y_train).predict_proba(X_test)
+```
+
+```{code-cell} ipython3
```
Comment (edublancas) [inline docs/classification/calibration.md:55]: organize this in 2-3 cells and add some explanation before each cell, otherwise it's too many steps to copy-paste each cell

**Review by edublancas — CHANGES_REQUESTED** [review]

File: CHANGELOG.md (hunk 1 above)
Comment (neelasha23) [reply CHANGELOG.md:6]: I think it happened because of a version bump before the review

**Comment by neelasha23** [comment]
Resolved the comments
@edublancas

File: src/sklearn_evaluation/plot/calibration.py [hunk 8]
```diff
@@ -29,16 +29,369 @@
...
+            classifier that produced the corresponding probability estimates in
+            `probabilities`. If ``None``, the names "Classifier 1", "Classifier 2",
+            etc. will be used.
+
+    cmap : string or :class:`matplotlib.colors.Colormap` instance, optional
+            Colormap used for plotting the projection. View Matplotlib Colormap
+            documentation for available options.
+            https://matplotlib.org/users/colormaps.html
+
+    name : string, optional
```
Comment (edublancas) [inline src/sklearn_evaluation/plot/calibration.py:129]: let's rename this to be consistent

```suggestion
    label : string, optional
```

File: src/sklearn_evaluation/plot/calibration.py [hunk 9]
```diff
@@ -29,16 +29,369 @@
...
+
+    @modify_exceptions
+    @SKLearnEvaluationLogger.log(feature="plot", action="calibration-curve-init")
+    def __init__(
+        self,
+        mean_predicted_value,
+        fraction_of_positives,
+        clf_names=None,
+        cmap="nipy_spectral",
+        name=None,
```
Comment (edublancas) [inline src/sklearn_evaluation/plot/calibration.py:153]:
```suggestion
        label=None,
```

**Review by edublancas — CHANGES_REQUESTED** [review]

**Comment by edublancas** [comment]
please rebase, there are some conflicts

File: src/sklearn_evaluation/plot/calibration.py (hunk 8 above)
Comment (neelasha23) [reply src/sklearn_evaluation/plot/calibration.py:129]: name is name of the plot. `label` in other plots refer to the curve labels (which in this case is `clf_names`).
Should I still rename?
@edublancas

Comment (neelasha23) [reply src/sklearn_evaluation/plot/calibration.py:129]: @edublancas ^^

Comment (edublancas) [reply src/sklearn_evaluation/plot/calibration.py:129]: I see. so let's keep the same signature `calibration_curve` to have backward compatibility. For the OO interface let's delete `name` (we can think of a way to allow users to customize the plot's title that's standard across different plots) and rename `clf_names` to `label`

**Review by edublancas — APPROVED** [review]
